File: thinking-orbs-mega/ports/ios/ThinkingOrbs/Sources/ThinkingOrbs/Profiles.swift
// Density profiles + the multiplier machinery that scales them. The base
// rows are the upstream `fine` profiles; each shipped preset (state ×
// size) applies count / radius multipliers on top, resolved once per
// (state, size) pair.

import Foundation

typealias ModeOpts = [String: Double]

// 2-D lattices (rings × dots-per-ring) come in pairs — each side takes
// √scale so the TOTAL dot count scales by `scale`; flat lists scale
// linearly. `iconD` sets the morph outline's sampling density.
private let countPairs: [(String, String)] = [
    ("latRings", "lonDensity"),
    ("rings", "lonDensity"),
    ("lanes", "segs")
]
private let countKeys = ["orbitN", "ghostN", "nodeN", "strandN", "signals"]
private let iconDensityKeys = ["iconD"]

// Every key that sets a dot's rendered radius — scaling all of them keeps
// a dot's near/far falloff intact while shrinking or growing the mark.
private let radiusKeys = [
    "rBase", "rDepth", "rActive", "rDot", "ghostR", "partR", "partRDepth",
    "nodeR", "nodeRDepth"
]

func scaleCounts(_ opts: ModeOpts, _ scale: Double) -> ModeOpts {
    var out = opts
    var done = Set<String>()
    let rt = scale.squareRoot()
    for (a, b) in countPairs {
        if let va = out[a], let vb = out[b], !done.contains(a), !done.contains(b) {
            out[a] = max(2, jsRound(va * rt))
            out[b] = max(2, jsRound(vb * rt))
            done.insert(a)
            done.insert(b)
        }
    }
    for k in countKeys {
        // Zero means the mode opted out of that layer entirely (ring has no
        // ghost sphere) — scaling must not resurrect it as a stray dot.
        if let v = out[k], v != 0, !done.contains(k) {
            out[k] = max(1, jsRound(v * scale))
        }
    }
    for k in iconDensityKeys {
        if let v = out[k] {
            out[k] = max(0.02, v * scale)
        }
    }
    return out
}

func scaleRadii(_ opts: ModeOpts, _ scale: Double) -> ModeOpts {
    var out = opts
    for k in radiusKeys {
        if let v = out[k] {
            out[k] = v * scale
        }
    }
    // remember the multiplier itself — spacing-derived radii use it,
    // since they aren't based on any single radius key
    out["rSizeMul"] = (out["rSizeMul"] ?? 1) * scale
    return out
}

/// Base (fine) profiles per mode, before preset multipliers.
let baseProfiles: [String: ModeOpts] = [
    "globe": [
        "latRings": 17,
        "lonDensity": 44,
        "rBase": 0.6,
        "rDepth": 1.7,
        "rBoost": 1.0,
        "inkFar": 0.62,
        "inkSpan": 0.54,
        "rsPow": 0.6,
        "rMin": 0.3
    ],
    "orbits": [
        "orbitN": 12,
        "ghostN": 40,
        "ghostR": 0.9,
        "ghostA": 0.5,
        "particles": 3,
        "partR": 1.2,
        "partRDepth": 1.6,
        "rsPow": 0.6,
        "rMin": 0.3
    ],
    "rubik": [
        "latRings": 15,
        "lonDensity": 40,
        "moveCount": 14,
        "rBase": 0.6,
        "rDepth": 1.7,
        "rActive": 0.3,
        "inkFar": 0.62,
        "inkSpan": 0.54,
        "rsPow": 0.6,
        "rMin": 0.3
    ],
    "wave": [
        "rings": 15,
        "lonDensity": 40,
        "rBase": 0.6,
        "rDepth": 1.7,
        "rsPow": 0.6,
        "rMin": 0.3
    ],
    "web": [
        "nodeN": 30,
        "thr": 0.72,
        "signals": 5,
        "nodeR": 1.4,
        "nodeRDepth": 1.8,
        "lineW": 0.8,
        "rsPow": 0.6,
        "rMin": 0.3
    ],
    "braid": [
        "strandN": 52,
        "turns": 3,
        "ghostN": 150,
        "rBase": 1.2,
        "rDepth": 1.8,
        "rsPow": 0.6,
        "rMin": 0.3
    ],
    "ribbon": [
        "lanes": 5,
        "segs": 88,
        "ghostN": 150,
        "rBase": 1.1,
        "rDepth": 1.7,
        "rsPow": 0.6,
        "rMin": 0.3
    ],
    // Ring shares ribbon's painter. `faceOn` cancels the camera tilt and
    // moves the undulation onto the radius; no ghost sphere sits behind it.
    "ring": [
        "lanes": 5,
        "segs": 88,
        "ghostN": 0,
        "faceOn": 1,
        "rBase": 1.1,
        "rDepth": 1.7,
        "rsPow": 0.6,
        "rMin": 0.3
    ],
    "morph": [
        "rDot": 0.021,
        "iconD": 1,
        "rMin": 0.25
    ]
]

File: thinking-orbs-mega/ports/svelte/src/lib/engine/profiles.ts
export interface ModeOpts {
	[key: string]: number | undefined;
}

const COUNT_PAIRS: ReadonlyArray<readonly [string, string]> = [
	["latRings", "lonDensity"],
	["rings", "lonDensity"],
	["lanes", "segs"],
];
const COUNT_KEYS = ["orbitN", "ghostN", "nodeN", "strandN", "signals"] as const;
const ICON_DENSITY_KEYS = ["iconD"] as const;
const RADIUS_KEYS = [
	"rBase",
	"rDepth",
	"rActive",
	"rDot",
	"ghostR",
	"partR",
	"partRDepth",
	"nodeR",
	"nodeRDepth",
] as const;

export function scaleCounts(opts: ModeOpts, scale: number): ModeOpts {
	const out: ModeOpts = { ...opts };
	const done = new Set<string>();
	const rt = Math.sqrt(scale);
	for (const [a, b] of COUNT_PAIRS) {
		const va = out[a];
		const vb = out[b];
		if (va != null && vb != null && !done.has(a) && !done.has(b)) {
			out[a] = Math.max(2, Math.round(va * rt));
			out[b] = Math.max(2, Math.round(vb * rt));
			done.add(a);
			done.add(b);
		}
	}
	for (const key of COUNT_KEYS) {
		const value = out[key];
		if (value != null && value !== 0 && !done.has(key))
			out[key] = Math.max(1, Math.round(value * scale));
	}
	for (const key of ICON_DENSITY_KEYS) {
		const value = out[key];
		if (value != null) out[key] = Math.max(0.02, value * scale);
	}
	return out;
}

export function scaleRadii(opts: ModeOpts, scale: number): ModeOpts {
	const out: ModeOpts = { ...opts };
	for (const key of RADIUS_KEYS) {
		const value = out[key];
		if (value != null) out[key] = value * scale;
	}
	out.rSizeMul = (out.rSizeMul ?? 1) * scale;
	return out;
}

export const BASE_PROFILES: Record<string, ModeOpts> = {
	globe: {
		latRings: 17,
		lonDensity: 44,
		rBase: 0.6,
		rDepth: 1.7,
		rBoost: 1,
		inkFar: 0.62,
		inkSpan: 0.54,
		rsPow: 0.6,
		rMin: 0.3,
	},
	orbits: {
		orbitN: 12,
		ghostN: 40,
		ghostR: 0.9,
		ghostA: 0.5,
		particles: 3,
		partR: 1.2,
		partRDepth: 1.6,
		rsPow: 0.6,
		rMin: 0.3,
	},
	rubik: {
		latRings: 15,
		lonDensity: 40,
		moveCount: 14,
		rBase: 0.6,
		rDepth: 1.7,
		rActive: 0.3,
		inkFar: 0.62,
		inkSpan: 0.54,
		rsPow: 0.6,
		rMin: 0.3,
	},
	wave: {
		rings: 15,
		lonDensity: 40,
		rBase: 0.6,
		rDepth: 1.7,
		rsPow: 0.6,
		rMin: 0.3,
	},
	web: {
		nodeN: 30,
		thr: 0.72,
		signals: 5,
		nodeR: 1.4,
		nodeRDepth: 1.8,
		lineW: 0.8,
		rsPow: 0.6,
		rMin: 0.3,
	},
	braid: {
		strandN: 52,
		turns: 3,
		ghostN: 150,
		rBase: 1.2,
		rDepth: 1.8,
		rsPow: 0.6,
		rMin: 0.3,
	},
	ribbon: {
		lanes: 5,
		segs: 88,
		ghostN: 150,
		rBase: 1.1,
		rDepth: 1.7,
		rsPow: 0.6,
		rMin: 0.3,
	},
	ring: {
		lanes: 5,
		segs: 88,
		ghostN: 0,
		faceOn: 1,
		rBase: 1.1,
		rDepth: 1.7,
		rsPow: 0.6,
		rMin: 0.3,
	},
	morph: {
		rDot: 0.021,
		iconD: 1,
		rMin: 0.25,
	},
};
